collector/src/main/java/io/prometheus/jmx/JmxScraper.java

```diff
@@ -108,17 +89,27 @@ public void doScrape() throws Exception {
         }
         try {
             // Query MBean names, see #89 for reasons queryMBeans() is used instead of queryNames()
-            Set<ObjectInstance> mBeanNames = new HashSet();
+            Set<ObjectName> mBeanNames = new HashSet<ObjectName>();
             for (ObjectName name : whitelistObjectNames) {
-                mBeanNames.addAll(beanConn.queryMBeans(name, null));
+                for (ObjectInstance instance : beanConn.queryMBeans(name, null)) {
+                    mBeanNames.add(instance.getObjectName());
+                }
             }
+
+            // Now that we have the whitelisted mBeans, remove any old ones from the cache:
+            jmxMBeanPropertyCache.onlyKeepMBeans(mBeanNames);
```

Comment:
Why not do this after the the blacklisting?

Reply:
That'll make it better, thanks!